### PR TITLE
fix: Refactor CMSMenu class for easier subclassing

### PR DESCRIPTION
**Pull request by fsbraun**
## Description

The `CMSMenu` class has been updated in #7807 for efficient and versioning-compatible menus.
This PR factors out two major changes to allow for future subclassing (such as asked for in https://github.com/django-cms/djangocms-versioning/issues/387):

* `CMSMenu.get_pagecontent_queryset`: returns the queryset to build the menu. The default implementation includes the admin-visible `latest_content`, e.g. not yet published pages, in edit and preview mode
* `CMSMenu.preview_endpoint`: returns a callable to get an endpoint for the menu entry. The default implementation returns the preview endpoint in edit or preview mode.

The PR does not change the logic of how `CMSMenu` works.

## Related resources

* https://github.com/django-cms/djangocms-versioning/issues/387
* #...

## Checklist

* [x] I have opened this pull request against ``develop-4``
* [ ] I have added or modified the tests when changing logic
* [x] I have followed [the conventional commits guidelines](https://www.conventionalcommits.org/) to add meaningful information into the changelog
* [x] I have read the [contribution guidelines ](https://github.com/django-cms/django-cms/blob/develop/CONTRIBUTING.rst) and I have joined #workgroup-pr-review on [Slack](https://www.django-cms.org/slack) to find a “pr review buddy” who is going to review my pull request.

### CODE REVIEW

**Comment by fsbraun** [comment]
@jrief Do you think this refactor will help you customize the menu system?

**Comment by jrief** [comment]
> @jrief Do you think this refactor will help you customize the menu system?

The bigger problem I have is that the menu renderer can't be replaced by a customized implementation, but that's a different story. I will merge this PR into my private branch and test it. Make take some days, though.

**Comment by fsbraun** [comment]
@jrief Is this still an issue? Or shall I close this?

File: cms/cms_menus.py [hunk 1]
```diff
@@ -198,6 +198,46 @@ class CMSMenu(Menu):
...
+        """
+        This method is used to get a callable based on the state of the toolbar which
+        is used to create the preview endpoint URL for each page content in the menu.
+        The callable must accept a page content object as a single parameter.
+
+        If the preview mode or edit mode is active in the toolbar, the callable will be set to the
+        get_object_preview_url function. Otherwise, it will be set to None leading to the fallback
+        of the get_absolute_url method of the page content.
+
+        :param toolbar: The toolbar object that contains the state information.
```
Comment (marksweb) [inline cms/cms_menus.py:231]: We've got inconsistent docstring formats here. 🤔

Comment (fsbraun) [reply cms/cms_menus.py:231]: Thanks, @marksweb ! I update the docstrings.

**Comment by fsbraun** [comment]
> The bigger problem I have is that the menu renderer can't be replaced by a customized implementation, but that's a different story. I will merge this PR into my private branch and test it. Make take some days, though.

@jrief Here's an example on how to replace the menu renderer by a (not very efficient) customized implementation:

https://github.com/django-cms/djangocms-versioning/blob/f31b5e0fb89ccccb486df4014eec53d6322aec5b/djangocms_versioning/cms_menus.py#L178-L181

**Comment by fsbraun** [comment]
I close this PR in favour of #7956

**Pull request closed without merging**